README.md
# servr

<!-- badges: start -->
[![R-CMD-check](https://github.com/yihui/servr/workflows/R-CMD-check/badge.svg)](https://github.com/yihui/servr/actions)
[![CRAN release](https://www.r-pkg.org/badges/version/servr)](https://cran.r-project.org/package=servr)
<!-- badges: end -->


A simple HTTP server to serve files under a given directory based on the
[**httpuv**](https://cran.r-project.org/package=httpuv) package.

You can install this package from
[CRAN](https://cran.r-project.org/package=servr) (stable version) or
[XRAN](https://xran.yihui.org) (development version):

```r
install.packages('servr')  # stable version; use a CRAN mirror, or
install.packages('servr', repos = 'https://xran.yihui.org')  # devel version
```

This package is licensed under GPL.

## Serve static files

To some degree, this package is like `python -m SimpleHTTPServer` or `python -m
http.server`. It may be used to serve:

- R Markdown applications
    - [**bookdown**](https://github.com/rstudio/bookdown) books
    - [**blogdown**](https://github.com/rstudio/blogdown) websites
    - [**xaringan**](https://github.com/yihui/xaringan) presentations
- [**googleVis**](https://cran.r-project.org/package=googleVis) charts (to avoid
  the Flash security issue)
- D3 charts which need to load JSON from local files (see also the
  [**animint**](https://github.com/tdhock/animint) package), and
  [**samatha**](https://github.com/DASpringate/samatha), etc

You can either run `servr::httd()` in an interactive R session, or run from
command line:

```bash
# default: port 4321, do not launch browser
Rscript -e "servr::httd()"

# open a web browser
Rscript -e "servr::httd()" -b

# listen on port 4000
Rscript -e "servr::httd()" -p4000

# pass arguments to the httd() function
Rscript -e "servr::httd(,4000,TRUE)"
```

There is also a shell script under `system.file('bin', package = 'servr')`;
if it is added to `PATH`, you can simply run

```bash
servr  # serve the current directory
servr -b  # launch the browser
servr -b -p4000  # change port to 4000
```

## Serve and watch a directory

Similar to `httd()`, the function `httw()` can both serve and watch a directory.
If you are viewing an HTML file in the browser, it will be automatically
refreshed whenever there are any changes in the directory (e.g. you added,
deleted, or modified certain files in the directory).

## Serve dynamic documents

Besides `httd()`, there are functions `jekyll()`, `rmdv1()`, and `rmdv2()` in
this package to serve HTML files generated from R Markdown documents (via
[**knitr**](https://yihui.org/knitr/) or
[**rmarkdown**](https://rmarkdown.rstudio.com)). R Markdown documents can be
automatically re-compiled when their HTML output files are older than the
corresponding source files, and HTML pages in the web browser can be
automatically refreshed accordingly, so you can focus on writing R Markdown
documents, and results will be updated on the fly in the web browser. This is
even more useful when you write R Markdown documents in the RStudio IDE, because
the HTML files are displayed in the RStudio viewer pane, and you can put the
source document and its output side by side.

![Jekyll with servr and knitr](http://i.imgur.com/gKVGhiP.png)

## Serve package vignettes

The function `vign()` can be used to serve R Markdown/HTML package vignettes.
The HTML output files are generated and displayed in the web browser so you can
preview the vignettes, and they will be cleaned up after they are loaded in the
web browser to make sure your source package is clean.

## Daemonized server

All server functions can be used in the daemonized mode, i.e., they can be
non-blocking in the R session, which allows you to continue working in the R
console after the server is launched. This mode can be set via the argument
`daemon = TRUE` in most server functions. See `?server_config` for more
information.


NEWS
		 CHANGES IN servr VERSION 0.22


		 CHANGES IN servr VERSION 0.21

NEW FEATURES

- Added a new argument `hosturl` to `server_config()`.

- Added a new argument `filter` to `httw()` to allow users to filter file paths on the watch list (thanks, @ARawles, #51).

		 CHANGES IN servr VERSION 0.20

MINOR CHANGES

- `server_config()` will add a leading `/` to `baseurl` if it has not already been included.

		 CHANGES IN servr VERSION 0.19

MINOR CHANGES

- `server_config()` also returns the `daemon` value in the list.

		 CHANGES IN servr VERSION 0.18

BUG FIXES

- Fixed the malformed Accept-Ranges header (thanks, @JasonPunyon, #47).

MINOR CHANGES

- `servr::random_port()` looks for an available port quietly via `httpuv::startServer(quiet = TRUE)`. This requires **httpuv** >= v1.5.2.

		 CHANGES IN servr VERSION 0.17

NEW FEATURES

- Added an argument `open` to `servr::browse_last()` so users can decide whether to reopen the lastly browsed page.

		 CHANGES IN servr VERSION 0.16

BUG FIXES

- Refinements to HTTP range request responses. Open-ended range requests (including "Range: bytes=0-") should now be correctly handled (thanks, @raymondben, #41).

		 CHANGES IN servr VERSION 0.15

BUG FIXES

- When the environment variable `R_SERVR_PORT` is set, `server_config()` may issue a superfluous warning "createTcpServer: address already in use" due to an unnecessary call to `random_port()` (thanks, @itcarroll, #39).

		 CHANGES IN servr VERSION 0.14

NEW FEATURES

- The argument `host` in `server_config()` can be configured through the global option `servr.host` now, e.g., `options(servr.host = '0.0.0.0')`.

		 CHANGES IN servr VERSION 0.13

NEW FEATURES

- Added a `verbose` argument to `server_config()`.

- The `interval` argument of `server_config()` can be set via the global option `servr.interval` now. For example, `options(servr.interval = 0.5)`.

MAJOR CHANGES

- Server functions such as `httd()` now return the config object created by `server_config()` instead of the server handle returned by `httpuv::startServer()`. The config object contains various information about the server and methods to start/stop the server.

		 CHANGES IN servr VERSION 0.12

NEW FEATURES

- Exported the function `random_port()` to return an available random TCP port.

MAJOR CHANGES

- `daemon_list()` returns server IDs instead of handles, so that it will work
  with httpuv >= v1.4.5.1 (rstudio/blogdown#365).

		 CHANGES IN servr VERSION 0.11

NEW FEATURES

- The default value of the `port` argument of `server_config()` can be set via
  the environment variable `R_SERVR_PORT`. If the environment variable does not
  exist, the global option `servr.port` will be used if set, e.g.,
  `options(servr.port = 4322)`. See `?servr::server_config` for details.

		 CHANGES IN servr VERSION 0.10

NEW FEATURES

- Added a function `browse_last()` to reopen the last browsed page.

MAJOR CHANGES

- The `daemon` argument in `server_config()` now defaults to `interactive()`, i.e., **servr** starts a daemonized server that does not block your interactive R session by default.

		 CHANGES IN servr VERSION 0.9

BUG FIXES

- The web browser may be opened too early (before the server is ready) (originally reported at rstudio/rstudio#2475).

		 CHANGES IN servr VERSION 0.8

NEW FEATURES

- added another implementation of the daemonized server based on the later package, since the previous implmentation based on `httpuv::startDaemonizedServer()` could crash the R session on Windows

		 CHANGES IN servr VERSION 0.7

NEW FEATURES

- added a new argument `watch` to `servr::httw()`

- exported the function `server_config()`

BUG FIXES

- files/directories that contain multibyte characters in path names cannot be served correctly (thanks, Hao Peng)

		 CHANGES IN servr VERSION 0.6

NEW FEATURES

- on 404 (page not found), 404.html will be displayed if it exists under the
  root directory

- improved the support for HTTP Range requests, e.g. servr can correctly serve
  MP4 videos now in major browsers including Safari

BUG FIXES

- servr should decode requested paths before reading them (https://github.com/rstudio/blogdown/issues/85)

		 CHANGES IN servr VERSION 0.5

NEW FEATURES

- added an argument `initpath` to `server_config()` so you can open a specific
  path initially in the web browser

- the `daemon` argument of `server_config()` takes the default value from the
  global option `getOption('servr.daemon')` now, e.g., you can set
  `options(servr.daemon = TRUE)` so that the daemonized server is always used

		 CHANGES IN servr VERSION 0.4.1

NEW FEATURES

- excluded ports considered unsafe by Chrome http://superuser.com/a/188070 when
  selecting a random port automatically

BUG FIXES

- fixed a bug in 301 redirection when serving a directory without the trailing
  slash

		 CHANGES IN servr VERSION 0.4

NEW FEATURES

- you can disable websocket listening on an HTML page using a special HTML
  comment `<!-- DISABLE-SERVR-WEBSOCKET -->` when servr is serving and watching
  a directory, so that this page will not communicate with R (e.g. when it is
  updated, R will not send signals to refresh it) (thanks @hafen, #25)

- a random TCP port will be used if the port 4321 is not available

		 CHANGES IN servr VERSION 0.3

NEW FEATURES

- added a function `httw()` to watch for changes under a directory and refresh
  an HTML page automatically (if it is being viewed in the browser) when any
  files are modified

- servr accepts HTTP Range requests now (thanks, @rekado, #21)

BUG FIXES

- servr did not work with RStudio Server (#20)

		 CHANGES IN servr VERSION 0.2

NEW FEATURES

- when running inside RStudio, the RStudio web browser will be used if
  available (requires RStudio >= 0.98.439)

- added three server functions jekyll(), rmdv1(), and rmdv2() to serve Jekyll
  websites, R Markdown v1, and R Markdown v2 documents, respectively

- added a server function vign() to serve R Markdown/HTML package vignettes


R/static.R
#' Serve static files under a directory
#'
#' If there is an \file{index.html} under this directory, it will be displayed;
#' otherwise the list of files is displayed, with links on their names. After we
#' run this function, we can go to \samp{http://localhost:port} to browse the
#' web pages either created from R or read from HTML files.
#'
#' \code{httd()} is a pure static server, and \code{httw()} is similar but
#' watches for changes under the directory: if an HTML file is being viewed in
#' the browser, and any files are modified under the directory, the HTML page
#' will be automatically refreshed.
#' @inheritParams server_config
#' @param ... server configurations passed to \code{\link{server_config}()}
#' @export
#' @references \url{https://github.com/yihui/servr}
#' @examples #' see https://github.com/yihui/servr for command line usage
#' # or run inside an R session
#' if (interactive()) servr::httd()
httd = function(dir = '.', ...) {
  dir = normalizePath(dir, mustWork = TRUE)
  if (dir != '.') {
    owd = setwd(dir); on.exit(setwd(owd))
  }
  res = server_config(dir, ...)
  app = list(call = serve_dir(dir))
  res$start_server(app)
  invisible(res)
}

#' @param watch a directory under which \code{httw()} is to watch for changes;
#'   if it is a relative path, it is relative to the \code{dir} argument
#' @param pattern a regular expression passed to \code{\link{list.files}()} to
#'   determine the files to watch
#' @param all_files whether to watch all files including the hidden files
#' @param filter a function to filter the file paths returned from
#'   \code{list.files()} (e.g., you can exclude certain files from the watch
#'   list)
#' @param handler a function to be called every time any files are changed or
#'   added under the directory; its argument is a character vector of the
#'   filenames of the files modified or added
#' @rdname httd
#' @export
httw = function(
  dir = '.', watch = '.', pattern = NULL, all_files = FALSE, filter = NULL,
  handler = NULL, ...
) {
  dynamic_site(dir, ..., build = watch_dir(
    watch, pattern = pattern, all_files = all_files, filter = filter, handler = handler
  ))
}

watch_dir = function(
  dir = '.', pattern = NULL, all_files = FALSE, filter = NULL, handler = NULL
) {
  cwd = getwd()
  mtime = function(dir) {
    owd = setwd(cwd); on.exit(setwd(owd), add = TRUE)
    info = file.info(list.files(
      dir, pattern, all.files = all_files, full.names = TRUE, recursive = TRUE,
      no.. = TRUE
    ))[, 'mtime', drop = FALSE]
    if (is.function(filter)) info = info[filter(rownames(info)), , drop = FALSE]
    rownames(info) = gsub('^[.]/', '', rownames(info))
    info
  }
  info = mtime(dir)
  function(...) {
    info2 = mtime(dir)
    changed = !identical(info, info2)
    if (changed) {
      if (is.function(handler)) {
        f1 = rownames(info)
        f2 = rownames(info2)
        f3 = setdiff(f2, f1)    # new files
        f4 = intersect(f1, f2)  # old files
        f5 = f4[info[f4, 1] != info2[f4, 1]]  # modified files
        info <<- info2
        handler(c(f3, na.omit(f5)))
        info2 = mtime(dir)
      }
      info <<- info2
    }
    changed
  }
}

#' Server configurations
#'
#' The server functions in this package are configured through this function.
#' @param dir The root directory to serve.
#' @param port The TCP port number. If it is not explicitly set, the default
#'   value will be looked up in this order: First, the command line argument of
#'   the form \code{-pNNNN} (N is a digit from 0 to 9). If it was passed to R
#'   when R was started, \code{NNNN} will be used as the port number. Second,
#'   the environment variable \code{R_SERVR_PORT}. Third, the global option
#'   \code{servr.port} (e.g., \code{options(servr.port = 4322)}). If none of
#'   these command-line arguments, variables, or options were set, the default
#'   port will be \code{4321}. If this port is not available, a random available
#'   port will be used.
#' @param browser Whether to launch the default web browser. By default, it is
#'   \code{TRUE} if the R session is \code{\link{interactive}()}, or when a
#'   command line argument \code{-b} was passed to R (see
#'   \code{\link{commandArgs}()}). N.B. the RStudio viewer is used as the web
#'   browser if available.
#' @param daemon Whether to launch a daemonized server (the server does not
#'   block the current R session) or a blocking server. By default, it is the
#'   global option \code{getOption('servr.daemon')} (e.g., you can set
#'   \code{options(servr.daemon = TRUE)}); if this option was not set,
#'   \code{daemon = TRUE} if a command line argument \code{-d} was passed to R
#'   (through \command{Rscript}), or the server is running in an interactive R
#'   session.
#' @param interval The time interval used to check if an HTML page needs to be
#'   rebuilt (by default, it is checked every second).
#' @param baseurl The base URL (the full URL will be
#'   \code{http://host:port/baseurl}).
#' @param initpath The initial path in the URL (e.g. you can open a specific
#'   HTML file initially).
#' @param hosturl A function that takes the host address and returns a character
#'   string to be used in the URL, e.g., \code{function(host) { if (host ==
#'   '127.0.0.1') 'localhost' else host}} to convert \code{127.0.0.1} to
#'   \code{localhost} in the URL.
#' @param verbose Whether to print messages when launching the server.
#' @inheritParams httpuv::startServer
#' @export
#' @return A list of configuration information of the form \code{list(host,
#'   port, start_server = function(app) {}, ...)}.
server_config = function(
  dir = '.', host = getOption('servr.host', '127.0.0.1'), port, browser, daemon,
  interval = getOption('servr.interval', 1), baseurl = '',
  initpath = '', hosturl = identity, verbose = TRUE
) {
  cargs = commandArgs(TRUE)
  if (missing(browser)) browser = interactive() || '-b' %in% cargs || is_rstudio()
  if (missing(port)) port = if (length(port <- grep('^-p[0-9]{4,}$', cargs, value = TRUE)) == 1) {
    sub('^-p', '', port)
  } else {
    port = Sys.getenv('R_SERVR_PORT', NA)
    if (is.na(port)) getOption('servr.port', random_port()) else port
  }
  port = as.integer(port)
  if (missing(daemon)) daemon = getOption('servr.daemon', ('-d' %in% cargs) || interactive())
  # rstudio viewer cannot display a page served at 0.0.0.0; use 127.0.0.1 instead
  host2 = if (host == '0.0.0.0' && is_rstudio()) '127.0.0.1' else host
  url = sprintf('http://%s:%d', hosturl(host2), port)
  baseurl = gsub('^/+', '', baseurl)
  if (baseurl != '') url = paste0(url, '/', baseurl)
  url = paste0(url, if (initpath != '' && !grepl('^/', initpath)) '/', initpath)
  browsed = FALSE
  servrEnv$browse = browse = function(reopen = FALSE) {
    if (browsed && !reopen) return(invisible(url))
    if (browser || reopen) browseURL(url, browser = get_browser())
    browsed <<- TRUE
    if (verbose && !reopen) message('Serving the directory ', dir, ' at ', url)
  }
  server = NULL
  list(
    host = host, port = port, interval = interval, url = url, daemon = daemon,
    start_server = function(app) {
      id = startServer(host, port, app)
      if (verbose && daemon) daemon_hint(id); browse()
      server <<- id
      if (!daemon) while (TRUE) {
        httpuv::service(); Sys.sleep(0.001)
      }
      invisible(id)
    },
    stop_server = function() {
      if (is.null(server)) stop('The server has not been started yet.')
      stopServer(server)
    },
    browse = browse
  )
}

serve_dir = function(dir = '.') function(req) {
  owd = setwd(dir); on.exit(setwd(owd), add = TRUE)
  path = decode_path(req)
  status = 200L

  if (grepl('^/', path)) {
    path = paste('.', path, sep = '')  # the requested file
  } else if (path == '') path = '.'
  body = if (file_test('-d', path)) {
    # ensure a trailing slash if the requested dir does not have one
    if (path != '.' && !grepl('/$', path)) return(list(
      status = 301L, body = '', headers = list(
        'Location' = sprintf('%s/', req$PATH_INFO)
      )
    ))
    type = 'text/html'
    if (file.exists(idx <- file.path(path, 'index.html'))) {
      readLines(idx, warn = FALSE)
    } else {
      d = file.info(list.files(path, all.files = TRUE, full.names = TRUE))
      title = escape_html(path)
      html_doc(c(sprintf('<h1>Index of %s</h1>', title), fileinfo_table(d)),
               title = title)
    }
  } else {
    # use the custom 404.html only if the path looks like a directory or .html
    try_404 = function(path) {
      file.exists('404.html') && grepl('(/|[.]html)$', path, ignore.case = TRUE)
    }
    # FIXME: using 302 here because 404.html may contain relative paths, e.g. if
    # /foo/bar/hi.html gives 404, I cannot just read 404.html and display it,
    # because it will be treated as /foo/bar/404.html; if 404.html contains
    # paths like ./css/style.css, I don't know how to let the browser know that
    # it means /css/style.css instead of /foo/bar/css/style.css
    if (!file.exists(path))
      return(if (try_404(path)) list(
        status = 302L, body = '', headers = list('Location' = '/404.html')
      ) else list(
        status = 404L, headers = list('Content-Type' = 'text/plain'),
        body = paste2('Not found:', path)
      ))

    type = guess_type(path)
    range = req$HTTP_RANGE

    if (is.null(range)) {
      read_raw(path)
    } else {
      range = strsplit(range, split = "(=|-)")[[1]]
      b2 = as.numeric(range[2])
      if (length(range) == 2 && range[1] == "bytes") {
        # open-ended range request
        # e.g. Chrome sends the range reuest 'bytes=0-'
        # http://stackoverflow.com/a/18745164/559676
        range[3] = file_size(path) - 1
      }
      b3 = as.numeric(range[3])
      if (length(range) < 3 || (range[1] != "bytes") || (b2 >= b3))
        return(list(
          status = 416L, headers = list('Content-Type' = 'text/plain'),
          body = 'Requested range not satisfiable\r\n'
        ))

      status = 206L  # partial content
      # type may also need to be changed
      # e.g. to "multipart/byteranges" if multipart range support is added at a later date
      # or possibly to "application/octet-stream" for binary files

      con = file(path, open = "rb", raw = TRUE)
      on.exit(close(con))
      seek(con, where = b2, origin = "start")
      readBin(con, 'raw', b3 - b2 + 1)
    }
  }
  if (is.character(body) && length(body) > 1) body = paste2(body)
  list(
    status = status, body = body,
    headers = c(list('Content-Type' = type), if (status == 206L) list(
      'Content-Range' = paste0("bytes ", range[2], "-", range[3], "/", file_size(path))
      ),
      'Accept-Ranges' = 'bytes') # indicates that the server supports range requests
  )
}
